### DISCUSSION TITLE
ioc相关问题

### DISCUSSION THREAD

**Discussion by 158545614** [BYJRK/DotNet-Discussions#390] [C# 技术交流]
请教一个关于IOC问题，我一个类A注册成瞬态的,这个类有个string属性，这个类A会作为其他类（类B，类C，类D等等）构造函数的入参，我在从容器获取其他类的时候，需要指定类A的string属性值，这个时候其他类（类B，类C，类D等等）构造函数中本来通过自动注入的参数是不是都要手动的指定了。
谢谢各位

**Comment by 158545614** [comment]
public DBHelper(string connectionstring, DbType dbType)
{
    this.Connectionstring = connectionstring;
    this.DbType = dbType;
}
public static void AddDBHelperTransient(this IServiceCollection services)
{
    services.AddTransient<Func<string,DbType, DBHelper>>(serviceProvider => (connectionString,dbType) => new DBHelper(connectionString,dbType));
}

——————————————————————————————————————————————————
private  IProgress singletonProgress;
public DBHelper dBHelper;

public MesWindowViewModel(IProgress singletonProgress)
{
    this.singletonProgress = singletonProgress;
    
}


public static void AddMesWindowSingleton(this IServiceCollection services)
{
    services.AddSingleton(ServiceProvider=> new MesWindowViewModel(ServiceProvider.GetService<IProgress>())
    {
        dBHelper= ServiceProvider.GetService<Func<string,DbType,DBHelper>>()("123",DbType.Sqlite)
    });

    services.AddSingleton(ServiceProvider => new MesWindow
    {
        DataContext = ServiceProvider.GetService<MesWindowViewModel>()
    });

}

如上面的代码，我的MesWindowViewModel中的IProgress singletonProgress本来是自动注入的，现在加了一个dBHelper字段，这个字段在注入的时候需要指定connectionstring, DbType，导致原本自动注入的singletonProgress需要手动指定（即通过services.AddSingleton(ServiceProvider=> new MesWindowViewModel(ServiceProvider.GetService<IProgress>())这里手动指定），有没有办法让IProgress不需要手动指定，只要指定dBHelper。谢谢各位

**Comment by TRARS** [comment]
给MesWindowViewModel安排个空接口IMesWindowViewModel
```C#
    services.AddSingleton<IMesWindowViewModel, MesWindowViewModel>();
    services.AddSingleton(sp=> 
    {
        var mesWindowViewModel = (MesWindowViewModel)(sp.GetRequiredService<IMesWindowViewModel>());
        {
            mesWindowViewModel.dBHelper = = sp.GetService<Func<string,DbType,DBHelper>>().Invoke("123",DbType.Sqlite);
        }
    });
```

**Reply by 158545614** [reply]
大佬 这个代码没法运行  我把多余的=号去掉了

**Reply by 158545614** [reply]
![image](https://github.com/user-attachments/assets/1b676400-863b-489a-8ed3-ea051e15bc23)

**Reply by 158545614** [reply]
我改成services.AddSingleton<IMESWindowViewModel, MesWindowViewModel>();
services.AddSingleton(sp =>
{
    var mesWindowViewModel = (MesWindowViewModel)(sp.GetRequiredService<IMESWindowViewModel>());
    {
        mesWindowViewModel.dBHelper =  sp.GetService<Func<string, DbType, DBHelper>>().Invoke("123", DbType.Sqlite);
    }
    return mesWindowViewModel;
});  运行后MesWindowViewModel里的dbHelper为空

**Reply by TRARS** [reply]
return忘写了。dBHelper不清楚，看着不应该是null啊。如果是null那GetService后面Invoke应该直接就报错了

**Reply by 158545614** [reply]
可能是上次我写的有问题，感谢大佬的方案，可行，也给了我新的思路。谢谢

是否还有其他简单一点的方案，再指点一下
